### PR TITLE
zulip-bots: Add `BotHandler` Protocol.

### LINKED ISSUE #639: Make a base class or protocol for ExternalBotHandler that zulip/zulip’s EmbeddedBotHandler can subclass
zulip/zulip currently needs an [unchecked cast](https://github.com/zulip/zulip/blob/a054f57af68637ad8b96290b3e13b2a6d8a8a359/zerver/worker/queue_processors.py#L748) to convert a custom `EmbeddedBotHandler` class into an `ExternalBotHandler`. Since `EmbeddedBotHandler` is not a subclass of `ExternalBotHandler` (because it works by invoking `internal_send_message` rather than making API calls), this is unsafe, and it breaks in practice when we try to upgrade python-zulip-api to 0.7.1 (zulip/zulip#16511).

To fix this, we need a common base class or [Mypy protocol](https://mypy.readthedocs.io/en/stable/protocols.html#simple-user-defined-protocols), say `BotHandler`, that can be subclassed by both `ExternalBotHandler` and `EmbeddedBotHandler`. The `zulip_bots` functions must be relaxed to accept `BotHandler` instead of requiring `ExternalBotHandler`.

If the `StubBotHandler` class used in tests could subclass `BotHandler` too, I think this would also let us correct various `bot_handler: Any` types to `bot_handler: BotHandler`.

### PR DESCRIPTION
**Pull request by LoopThrough-i-j**
- zulip-bots: Add `CommonStorage` Protocol.
   -  The `CommonStorage` Protocol is created to add a
  Common type to all Storage Classes.
   -  The Storages used in `BotHandler`s are typed as `CommonStorage`

    Note: Protocol is imported from `typing_extensions` as `typing `doesn't
    provide Protocol for python <= 3.7.

- zulip-bots: Add `BotHandler` Protocol.

     The `BotHandler` Protocol has methods which allow
     any of the Current BotHandlers to use `BotHandler` Protocol
     as a mypy type.

    Fixes part of #639

- zulip-bots: Set `bot_handler` type to `BotHandler`.
    - Set `bot_handler` type to `BotHandler`.
    - Fix mypy issues in improperly typed variables and returns.

    Fixes #639

### CODE REVIEW

File: zulip_bots/zulip_bots/lib.py [hunk 1]
```diff
@@ -102,7 +102,74 @@ def __init__(self, name: str, email: str) -> None:
...
+    def upload_file(self, file: IO[Any]) -> Dict[str, Any]:
+        pass
+
+    def unique_reply(self) -> Dict[str, Any]:
+        pass
+
+    def unique_response(self) -> Dict[str, Any]:
+        pass
+
+    def ensure_unique_response(
```
Comment (neiljp) [inline zulip_bots/zulip_bots/lib.py:163]: Aren't these just test helpers? Similarly reset_transcript and maybe others?

File: zulip_bots/zulip_bots/lib.py [hunk 2]
```diff
@@ -102,7 +102,74 @@ def __init__(self, name: str, email: str) -> None:
...
+    ) -> Dict[str, Any]:
+        pass
+
+    def send_reply(
+        self,
+        message: Dict[str, Any],
+        response: str,
+        widget_content: Optional[str] = None
+    ) -> Dict[str, Any]:
+        pass
```
Comment (neiljp) [inline zulip_bots/zulip_bots/lib.py:136]: Should this be `...` ?

File: zulip_bots/zulip_bots/lib.py [hunk 3]
```diff
@@ -9,7 +9,7 @@
 
 
 from typing import Any, Optional, List, Dict, IO, Text
-
+from typing_extensions import Protocol
```
Comment (neiljp) [inline zulip_bots/zulip_bots/lib.py:12]: This module has not been added to dependencies to be installed?

**Review by neiljp — COMMENTED** [review]
@LoopThrough-i-j Some initial thoughts from a quick look. I think the methods could be one-liners in some cases with the ellipsis.

File: zulip_bots/zulip_bots/lib.py (hunk 3 above)
Comment (LoopThrough-i-j) [reply zulip_bots/zulip_bots/lib.py:12]: I don't think we need a separate installation for this library.

File: zulip_bots/zulip_bots/lib.py (hunk 1 above)
Comment (LoopThrough-i-j) [reply zulip_bots/zulip_bots/lib.py:163]: These are test helpers and they were added because there were a few failing tests when `bot_handler` was assigned the type `BotHandler` in the next commit.

File: zulip_bots/zulip_bots/lib.py (hunk 3 above)
Comment (neiljp) [reply zulip_bots/zulip_bots/lib.py:12]: I don't believe typing_extensions is included in python3.6?
I tested this in a venv and `zulip-run-bot` rightly says
```
ModuleNotFoundError: No module named 'typing_extensions'

File: zulip_bots/zulip_bots/lib.py (hunk 1 above)
Comment (LoopThrough-i-j) [reply zulip_bots/zulip_bots/lib.py:163]: Will still take a look at improving it if possible.

File: zulip_bots/zulip_bots/lib.py (hunk 3 above)
Comment (LoopThrough-i-j) [reply zulip_bots/zulip_bots/lib.py:12]: Will add that.

File: zulip_bots/zulip_bots/lib.py (hunk 1 above)
Comment (LoopThrough-i-j) [reply zulip_bots/zulip_bots/lib.py]: @neiljp yeah, I considered removing the helper functions which were not really necessary.

File: zulip_bots/zulip_bots/lib.py (hunk 2 above)
Comment (LoopThrough-i-j) [reply zulip_bots/zulip_bots/lib.py:133]: I am not sure about what you are saying! Do you mean I should use *args or **kwargs for this because of the EmbeddedBotHandler in Zulip?

Comment (LoopThrough-i-j) [reply zulip_bots/zulip_bots/lib.py:133]: I am not sure of what is `...`!! Do you mean these should be in 1 line?

Comment (neiljp) [reply zulip_bots/zulip_bots/lib.py:121]: https://mypy.readthedocs.io/en/stable/protocols.html#simple-user-defined-protocols

Comment (LoopThrough-i-j) [reply zulip_bots/zulip_bots/lib.py:121]: Ahhh, ok I understand now. Probably missed this part while going through the library.

Comment (LoopThrough-i-j) [reply zulip_bots/zulip_bots/lib.py]: Thanks, I resolved it.

**Comment by LoopThrough-i-j** [comment]
@neiljp thanks for pointing out some improvements. I had actually added inherited the `BotHandler` class and added some helper functions in the `BotHandler`, confused by the innumerable number of errors thrown by the bots after the 2nd commit, which I slowly debugged and figured out and then considered keeping the `BotHandler` the same.

**Review by neiljp — COMMENTED** [review]
@LoopThrough-i-j This looks tidier already:+1: I'd check the server bots for the methods in common (quit?), or their types (Any?).

File: zulip_bots/zulip_bots/lib.py [hunk 4]
```diff
@@ -102,6 +102,33 @@ def __init__(self, name: str, email: str) -> None:
         self.email = email
         self.mention = '@**' + name + '**'
 
+class BotHandler(Protocol):
+
+    email: str
+    full_name: str
+    storage: Any
```
[the remark this answers is not in the document]
Comment (LoopThrough-i-j) [reply zulip_bots/zulip_bots/lib.py:109]: Okay, am taking a look.

File: zulip_bots/zulip_bots/lib.py [hunk 5]
```diff
@@ -102,6 +102,33 @@ def __init__(self, name: str, email: str) -> None:
...
+
+    def identity(self) -> BotIdentity:
+        ...
```
[the remark this answers is not in the document]
Comment (LoopThrough-i-j) [reply zulip_bots/zulip_bots/lib.py:112]: Flake8 will throw errors :
```flake8    | zulip_bots/zulip_bots/lib.py:111:5: E704 multiple statements on one line (def)```

File: zulip_bots/zulip_bots/lib.py [hunk 6]
```diff
@@ -188,7 +215,7 @@ def update_message(self, message: Dict[str, Any]) -> Dict[str, Any]:
             self._rate_limit.show_error_and_exit()
         return self._client.update_message(message)
 
-    def get_config_info(self, bot_name: str, optional: Optional[bool] = False) -> Dict[str, Any]:
+    def get_config_info(self, bot_name: str, optional: bool = False) -> Dict[str, Any]:
```
[the remark this answers is not in the document]
Comment (LoopThrough-i-j) [reply zulip_bots/zulip_bots/lib.py:218]: Yeah, anyways Optional[bool] doesn't make sense as we are assigning it False by default. I will add it to the commit message.

File: zulip_bots/zulip_bots/lib.py [hunk 7]
```diff
@@ -102,6 +102,33 @@ def __init__(self, name: str, email: str) -> None:
...
+    def send_reply(self, message: Dict[str, Any], response: str, widget_content: Optional[str] = None) -> Dict[str, Any]:
+        ...
+
+    def update_message(self, message: Dict[str, Any]) -> Optional[Dict[str, Any]]:
+        ...
+
+    def get_config_info(self, bot_name: str, optional: bool = False) -> Dict[str, Any]:
+        ...
+
+    def quit(self, message: str = "") -> None:
```
[the remark this answers is not in the document]
Comment (LoopThrough-i-j) [reply zulip_bots/zulip_bots/lib.py:129]: Yeah, probably quit is also not required.

File: zulip_bots/zulip_bots/lib.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (neiljp) [reply zulip_bots/zulip_bots/lib.py:218]: The default value is independent of the type. I doubt the type is `Union[None, bool]`, but that's a separate point.

File: zulip_bots/zulip_bots/lib.py [hunk 8]
```diff
@@ -102,6 +102,33 @@ def __init__(self, name: str, email: str) -> None:
...
+    full_name: str
+    storage: Any
+
+    def identity(self) -> BotIdentity:
+        ...
+
+    def react(self, message: Dict[str, Any], emoji_name: str) -> Dict[str, Any]:
+        ...
+
+    def send_message(self, message: (Dict[str, Any])) -> Dict[str, Any]:
```
[the remark this answers is not in the document]
Comment (LoopThrough-i-j) [reply zulip_bots/zulip_bots/lib.py:117]: Ohh, it is because I copied the function from `ExternalBotHandler` and it had the same.

File: zulip_bots/zulip_bots/lib.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (LoopThrough-i-j) [reply zulip_bots/zulip_bots/lib.py:218]: okay will still take a look, tests pass though.

File: zulip_bots/zulip_bots/lib.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (LoopThrough-i-j) [reply zulip_bots/zulip_bots/lib.py:129]: getting this error when removing quit.
```
mypy      | zulip_bots/zulip_bots/bots/idonethis/idonethis.py:140: error: "BotHandler" has no attribute "quit"
```

File: zulip_bots/zulip_bots/lib.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (LoopThrough-i-j) [reply zulip_bots/zulip_bots/lib.py:218]: Changed it to the original.

File: zulip_bots/zulip_bots/lib.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (LoopThrough-i-j) [reply zulip_bots/zulip_bots/lib.py]: Decided to create a `CommonStorage` Protocol to solve this.

File: zulip_bots/zulip_bots/lib.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (neiljp) [reply zulip_bots/zulip_bots/lib.py]: I wasn't meaning it shouldn't be changed, but we can leave it for now, pending what others think.

File: zulip_bots/zulip_bots/lib.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (neiljp) [reply zulip_bots/zulip_bots/lib.py:141]: Did you check the server code? It does seem like it is there.

File: zulip_bots/setup.py [hunk 9]
```diff
@@ -65,6 +65,7 @@
         'html2text',
         'lxml',
         'BeautifulSoup4',
+        'typing_extensions',
```
Comment (neiljp) [inline zulip_bots/setup.py:68]: Is the change to requirements.txt necessary?
Do we need a specific/minimum version of typing_extensions?

File: zulip_bots/zulip_bots/test_lib.py [hunk 10]
```diff
@@ -25,7 +26,7 @@
 
 class StubBotHandler:
     def __init__(self) -> None:
-        self.storage = SimpleStorage()
+        self.storage = SimpleStorage()  # type: CommonStorage
```
Comment (neiljp) [inline zulip_bots/zulip_bots/test_lib.py:29]: This looks good, though I'm not sure if we need a separate protocol for this or if this will be a common implementation.
I would put this refactor into a separate commit.
We should be able to use an inline type annotation.

File: zulip_bots/zulip_bots/lib.py [hunk 11]
```diff
@@ -102,6 +102,45 @@ def __init__(self, name: str, email: str) -> None:
...
+
+    def react(self, message: Dict[str, Any], emoji_name: str) -> Dict[str, Any]:
+        ...
+
+    def send_message(self, message: Dict[str, Any]) -> Dict[str, Any]:
+        ...
+
+    # For ExternalBotHandler and StubBotHandler returns Dict[str, Any]
+    # Whereas EmbeddedBotHandler returns None
+    def send_reply(self, message: Dict[str, Any], response: str, widget_content: Optional[str] = None) -> Any:
```
Comment (neiljp) [inline zulip_bots/zulip_bots/lib.py:132]: There is also variation in other methods?

**Review by neiljp — COMMENTED** [review]
@LoopThrough-i-j This is looking better, though please check the server code for comparison if you've not already. Some of the errors you found are likely since the implementation is using a BotHandler and that doesn't supply the method, it's possible only an ExternalBotHandler does.

There seem to be quite a few deviations between the External and Embedded cases, so I'm not sure where to align to here, though since embedded bots presumably are upgraded with the server they could be adapted more easily?

@andersk I've given this some passes, but I'd appreciate your input based on the original motivation and how you see the two versions coming together.

File: zulip_bots/zulip_bots/lib.py (hunk 11 above)
Comment (andersk) [reply zulip_bots/zulip_bots/lib.py:132]: We should avoid `Any` wherever possible because it’s completely unchecked; mypy will silently allow you to put in one type and get out another. Consider one of

* `-> object` (see [`Any` vs. `object`](https://mypy.readthedocs.io/en/stable/dynamic_typing.html#any-vs-object));
* `-> Optional[Dict[str, Any]]`;
* `-> Dict[str, Any]` and add the return value in all cases;
* `-> None` and remove the return value in all cases.

Comment (LoopThrough-i-j) [reply zulip_bots/zulip_bots/lib.py:132]: @andersk 
If we use any of:
- `Optional[Dict[str, Any]]`;
- `Dict[str, Any]`;
- `None`

The method will start expecting the same return declaration in the actual class methods.
Tried Overloading as well but didn't work.

Thanks for the object vs Any, comparison. I think I should update it as an object.

File: zulip_bots/zulip_bots/test_lib.py (hunk 10 above)
Comment (LoopThrough-i-j) [reply zulip_bots/zulip_bots/test_lib.py:29]: The only way to provide a static type to storage is by using a Protocol.
Otherwise making it an object is fine as well, but I think it is a better way to go.

File: zulip_bots/setup.py (hunk 9 above)
Comment (LoopThrough-i-j) [reply zulip_bots/setup.py:68]: Yeah, they aren't, noticed it later.

File: zulip_bots/zulip_bots/lib.py (hunk 11 above)
Comment (andersk) [reply zulip_bots/zulip_bots/lib.py:132]: It’s perfectly [valid](https://mypy-play.net/?mypy=latest&python=3.9&gist=dd4e030f4136c7cd74d074e04fd2c8d8) to implement a protocol method returning `Optional[Dict[str, Any]]` with an implementation method returning `Dict[str, Any]` or `None`, as these are both subtypes of `Optional[Dict[str, Any]]`.

For the other two suggestions, you’ve left off the important part: “and add/remove the return value in all cases”. I think one of these is probably the right plan. Even though all four plans would be accepted by mypy, there’s no actual reason to a return a value in some implementations but not others. What’s the point if the caller can’t rely on it being there?

Avoid overloading; it’s also completely unchecked.

Comment (LoopThrough-i-j) [reply zulip_bots/zulip_bots/lib.py:132]: But I think we cannot use objects here because it says objects are not indexable, but Dicts are. So, I think Any is the only way to go. The same reason why we can't use `Optional[Dict[str, Any]]`, Optional will give it a possibility of None making it non-indexable as well.

Comment (andersk) [reply zulip_bots/zulip_bots/lib.py:132]: `None` doesn’t become any more indexable just because you hid it away in an `Any` type. Mypy won’t be able to find the error, but it’s still an error.

Comment (andersk) [reply zulip_bots/zulip_bots/lib.py:132]: (Can you reply with normal replies instead of editing your original comment? It’s confusing to try to have this discussion out of order.)

I don’t see why you’re saying “that is not true for a `Dict` return” when I just demonstrated that it is.

File: zulip_bots/zulip_bots/test_lib.py (hunk 10 above)
Comment (LoopThrough-i-j) [reply zulip_bots/zulip_bots/test_lib.py:29]: Pushed in a separate commit.

File: zulip_bots/zulip_bots/lib.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (LoopThrough-i-j) [reply zulip_bots/zulip_bots/lib.py:143]: yeah, they quit as well.

File: zulip_bots/zulip_bots/lib.py [hunk 12]
```diff
@@ -102,6 +102,47 @@ def __init__(self, name: str, email: str) -> None:
...
+        ...
+
+    def send_message(self, message: Dict[str, Any]) -> Dict[str, Any]:
+        ...
+
+    # ExternalBotHandler and StubBotHandler returns Dict[str, Any]
+    # Whereas EmbeddedBotHandler returns None
+    # Optional[Dict[str, Any]] or object cannot be used because there are occasions where
+    # function calls to send_reply strictly expects Dict[str, Any].
+    def send_reply(self, message: Dict[str, Any], response: str, widget_content: Optional[str] = None) -> Any:
```
Comment (andersk) [inline zulip_bots/zulip_bots/lib.py:134]: If there are callers that expect `Dict[str, Any]`, and there are implementation that provide `None`, then that is a bug. The correct resolution isn’t to reach for `Any` and prevent mypy from seeing that there is a bug. The correct resolution is to fix the bug. Either the implementations should be fixed to provide `Dict[str, Any]`, or the callers should be fixed to check for `None`.

File: zulip_bots/zulip_bots/lib.py (hunk 11 above)
Comment (LoopThrough-i-j) [reply zulip_bots/zulip_bots/lib.py]: I am sorry to edit my message, I thought of condensing everything in one message. Won't do that from now. 
[this](https://mypy-play.net/?mypy=latest&python=3.9&gist=dd4e030f4136c7cd74d074e04fd2c8d8) defines the error.

Comment (LoopThrough-i-j) [reply zulip_bots/zulip_bots/lib.py]: oops it didn't save

Comment (LoopThrough-i-j) [reply zulip_bots/zulip_bots/lib.py]: I am not sure how to share mypy playground:
The error is thrown when we run this:

try this out:
```
from typing import Any, Dict, Optional

from typing_extensions import Protocol

class P(Protocol):
    def foo(self) -> Optional[Dict[str, Any]]:
        ...

class A:
    def foo(self) -> Dict[str, Any]:
        return {"test": "test"}

class B:
    def foo(self) -> None:
        pass

new = A() # type: P
d = new.foo()
print(d["test"])
```

File: zulip_bots/zulip_bots/lib.py (hunk 12 above)
Comment (LoopThrough-i-j) [reply zulip_bots/zulip_bots/lib.py:134]: The `ExternalBotHandler` always returns a `Dict` [see this](https://github.com/zulip/python-zulip-api/blob/3887ad102e26efd1a9e3ed80e95a3ca4ad6600f8/zulip_bots/zulip_bots/lib.py#L161), but the `EmbeddedBotHandler` always returns `None` [see this](https://github.com/zulip/zulip/blob/829f9272d2c4299a0c0a37a09802248d8136c0a8/zerver/lib/bot_lib.py#L111). Our Protocol should be compatible to both I hope. If not then probably I understood it wrong.

Comment (LoopThrough-i-j) [reply zulip_bots/zulip_bots/lib.py:134]: SO I don't think it is a problem for caller's side

File: zulip_bots/zulip_bots/lib.py (hunk 11 above)
Comment (andersk) [reply zulip_bots/zulip_bots/lib.py]: Right: `A()` is a valid implementation of `P`, but `d["test"]` is not a valid use of `P`.

If you want `B()` to be a valid implementation, then `d["test"]` cannot be a valid use. If you want `d["test"]` to be a valid use, then `B()` cannot be a valid implementation. One or the other will need to be fixed.

Comment (LoopThrough-i-j) [reply zulip_bots/zulip_bots/lib.py]: Yes, that's my point. I have commented on the reason behind me choosing `Any` is in reply to your next comment.

Comment (andersk) [reply zulip_bots/zulip_bots/lib.py]: But this is not a reason to choose `Any`!

We are clearly failing to communicate somewhere, so let me try to simplify. Let’s take all the protocols and classes out of this and just look at some basic types.

Mypy rejects this program because it is wrong:

```python
d = None
print(d["test"])
```

Mypy accepts this program, but it is **still just as wrong**:

```python
d: Any = None
print(d["test"])
```

Using `Any` has **not fixed the bug**. It has only silenced mypy.

Our goal is not to silence mypy. Our goal is to fix the bug.

We can fix the bug by having the consumer check for `None`:

```python
d: Optional[Dict[str, str]] = None
if d is not None:
    print(d["test"])
```

or we can fix the bug by having the producer avoid producing `None`:

```python
d: Dict[str, str] = {"test": "yay"}
print(d["test"])
```

Comment (LoopThrough-i-j) [reply zulip_bots/zulip_bots/lib.py]: I thought we shouldn't put `Optional` because `bot_handler` was an instance of `StubBotHandler` which returns `Dict` strictly, but we have made `bot_handler` an instance of `BotHandler` which made the type `Optional[Dict]`.

will fix this error:
```
mypy      | zulip_bots/zulip_bots/bots/incrementor/incrementor.py:37: error: Value of type "object" is not indexable
```
by putting a check and type as Optional[Dict].

I think fix 1:
```python
d: Optional[Dict[str, str]] = None
if d is not None:
    print(d["test"])
```
is better, then we would ideally be able to send any kind of bot_handler to `bot_handler` parameter.

Thanks for helping soo much, and sorry for being a dumb kid.

**Comment by andersk** [comment]
In order to use this in zulip/zulip, we also need to adjust `zulip_bots.lib.extract_query_without_mention` to take `BotHandler` instead of `ExternalBotHandler`. There may be other such adjustments that make sense.

**Comment by LoopThrough-i-j** [comment]
> In order to use this in zulip/zulip, we also need to adjust `zulip_bots.lib.extract_query_without_mention` to take `BotHandler` instead of `ExternalBotHandler`. There may be other such adjustments that make sense.

Thanks, I didn't know. Will get things fixed soon.

**Review by andersk — APPROVED** [review]
I renamed `CommonStorage` to `BotStorage`. Otherwise looks good.
